ee/middleware/upstreamoauth/middleware.go

```diff
@@ -0,0 +1,101 @@
+package upstreamoauth
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/TykTechnologies/tyk/storage"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/TykTechnologies/tyk/header"
+	"github.com/TykTechnologies/tyk/internal/httputil"
+	"github.com/TykTechnologies/tyk/internal/model"
+)
+
+// Middleware implements upstream basic auth middleware.
+type Middleware struct {
+	Spec model.MergedAPI
+	Gw   Gateway
+
+	Base                            BaseMiddleware
+	clientCredentialsStorageHandler *storage.RedisCluster
+	passwordStorageHandler          *storage.RedisCluster
+}
+
+// Middleware implements model.Middleware.
+var _ model.Middleware = &Middleware{}
+
+// NewMiddleware returns a new instance of Middleware.
+func NewMiddleware(gw Gateway, mw BaseMiddleware, spec model.MergedAPI, ccStorageHandler *storage.RedisCluster, pwStorageHandler *storage.RedisCluster) *Middleware {
+	return &Middleware{
+		Base:                            mw,
+		Gw:                              gw,
+		Spec:                            spec,
+		clientCredentialsStorageHandler: ccStorageHandler,
+		passwordStorageHandler:          pwStorageHandler,
+	}
+}
+
+// Logger returns a logger with middleware filled out.
+func (m *Middleware) Logger() *logrus.Entry {
+	return m.Base.Logger().WithField("mw", m.Name())
+}
+
+// Name returns the name for the middleware.
+func (m *Middleware) Name() string {
+	return MiddlewareName
+}
+
+// EnabledForSpec checks if streaming is enabled on the config.
+func (m *Middleware) EnabledForSpec() bool {
+	if !m.Spec.UpstreamAuth.IsEnabled() {
+		return false
+	}
+
+	if !m.Spec.UpstreamAuth.OAuth.Enabled {
+		return false
+	}
+
+	return true
+}
+
+// Init initializes the middleware.
+func (m *Middleware) Init() {
+	m.Logger().Debug("Initializing Upstream basic auth Middleware")
+}
+
+// ProcessRequest will handle upstream basic auth.
+func (m *Middleware) ProcessRequest(_ http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	oauthConfig := m.Spec.UpstreamAuth.OAuth
+
+	upstreamOAuthProvider := Provider{
+		HeaderName: header.Authorization,
+	}
+
+	provider, err := getOAuthHeaderProvider(oauthConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get OAuth header provider: %w", err), http.StatusInternalServerError
+	}
+
+	payload, err := provider.getOAuthToken(r, m)
+	if err != nil {
+		return fmt.Errorf("failed to get OAuth token: %w", err), http.StatusInternalServerError
+	}
+
+	upstreamOAuthProvider.AuthValue = payload
```

Comment:
This function isn't structured properly. Provide a constructor for Provider passing in required things.
Why does resetting the header run twice? Should delete L87-89 if the intent was to do it only for some providers.
The name provider here repeats, quite confusing to read.